fpgaconvnet/models/layers/ConvolutionLayer3D.py

```diff
@@ -20,93 +21,70 @@
-
-        # save bias flag
-        self.has_bias = has_bias
-
-        # init variables
-        self._kernel_rows = kernel_rows
-        self._kernel_cols = kernel_cols
-        self._kernel_depth = kernel_depth
-        self._stride_rows = stride_rows
-        self._stride_cols = stride_cols
-        self._stride_depth = stride_depth
-        self._pad_top = pad_top
-        self._pad_right = pad_right
-        self._pad_front = pad_front
-        self._pad_bottom = pad_bottom
-        self._pad_left = pad_left
-        self._pad_back = pad_back
-        self._groups = groups
-        self._coarse_group = coarse_group
-        self._fine = fine
-        self._filters = filters
+    filters: int
+    kernel_rows: int
+    kernel_cols: int
+    kernel_depth: int
+    coarse_in: int = 1
+    coarse_out: int = 1
+    coarse_group: int = 1
+    stride_rows: int = 2
+    stride_cols: int = 2
+    stride_depth: int = 2
+    groups: int = 1
+    pad_top: int = 0
+    pad_right: int = 0
+    pad_front: int = 0
+    pad_bottom: int = 0
+    pad_left: int = 0
+    pad_back: int = 0
+    fine: int  = 1
+    input_t: FixedPoint = field(default_factory=lambda: FixedPoint(16,8), init=True)
+    output_t: FixedPoint = field(default_factory=lambda: FixedPoint(16,8), init=True)
+    weight_t: FixedPoint = field(default_factory=lambda: FixedPoint(16,8), init=True)
+    acc_t: FixedPoint = field(default_factory=lambda: FixedPoint(32,16), init=True)
+    has_bias: int = 0 # default to no bias for old configs
+    sparsity: list = field(default_factory=lambda: [], init=True)
+    block_floating_point: bool = False
+    backend: str = "chisel" # default to no bias for old configs
+    regression_model: str = "linear_regression"
+    stream_weights: int = 0
+    stream_inputs: list = field(default_factory=lambda: [0], init=True)
+
+    def __post_init__(self):
+
+        # call parent post init
+        super().__post_init__()
 
         # check if the layer is depthwise
-        self.depthwise = (groups == channels) and (groups == filters)
+        self.depthwise = (self.groups == self.channels) and (self.groups == self.filters)
+        self.pointwise = np.prod(self.kernel_size) == 1
 
-        # backend flag
-        assert backend in ["hls", "chisel"], f"{backend} is an invalid backend"
-        self.backend = backend
+        # save sparsity
+        if len(self.sparsity) > 0:
+            # reject if pointwise or low sparsity
+            if self.pointwise or np.mean(self.sparsity) < 0.1:
+                self.sparsity = []
+        # self.sparsity = sparsity
 
         # weights buffering flag
         if self.backend == "hls":
             self.double_buffered = False
-            self.stream_weights = False
+            self.stream_weights = 0
             self.data_packing = False
+            self.use_uram = False
         elif self.backend == "chisel":
             self.double_buffered = False
-            self.stream_weights = False
```

Comment:
same here

Reply:
@AlexMontgomerie Any idea about this one?